File: SatinalmaPro.Shared/Procurement/Detail/PurchaseRequestDetailPresenter.cs
using SatinalmaPro.Shared.Helpers;
using SatinalmaPro.Shared.Models;

namespace SatinalmaPro.Shared.Procurement.Detail;

/// <summary>
/// Talep detay ekranı — durum, öncelik ve role göre buton görünürlüğü.
/// Masaüstü ve Android ortak presenter.
/// </summary>
public static class PurchaseRequestDetailPresenter
{
    private static readonly IReadOnlyDictionary<PurchaseRequestDetailAction, string> DefaultLabels =
        new Dictionary<PurchaseRequestDetailAction, string>
        {
            [PurchaseRequestDetailAction.DirectApprove] = "Direkt Onay Ver",
            [PurchaseRequestDetailAction.RejectRequest] = "Talebi Reddet",
            [PurchaseRequestDetailAction.StartQuoteProcess] = "Teklif Sürecini Başlat",
            [PurchaseRequestDetailAction.ApproveQuote] = "Bu Firmayı Onayla",
            [PurchaseRequestDetailAction.RejectEntireRequest] = "Talebi Komple Reddet",
            [PurchaseRequestDetailAction.SendQuotesForRevision] = "Teklifleri Revizeye Gönder"
        };

    public static PurchaseRequestDetailScreen ResolveScreen(SatinalmaTalep talep)
    {
        var status = ProcurementStatusResolver.Resolve(talep);
        return status.Equals(ProcurementStatus.ManagementQuoteReview, StringComparison.OrdinalIgnoreCase)
            ? PurchaseRequestDetailScreen.ManagementQuoteReview
            : PurchaseRequestDetailScreen.ManagementSubmittedReview;
    }

    public static string ResolvePriority(SatinalmaTalep talep)
    {
        if (!string.IsNullOrWhiteSpace(talep.Priority)
            && !talep.Priority.Equals(ProcurementPriority.Normal, StringComparison.OrdinalIgnoreCase))
            return ProcurementPriority.Normalize(talep.Priority);

        return ProcurementPriority.FromRequestType(talep.TalepTuru);
    }

    public static PurchaseRequestDetailUiState BuildUiState(
        SatinalmaTalep talep,
        string? role,
        PurchaseRequestDetailScreen? screen = null)
    {
        var resolvedScreen = screen ?? ResolveScreen(talep);
        var status = ProcurementStatusResolver.Resolve(talep);
        var priority = ResolvePriority(talep);
        var canManage = CanManagementDecide(role);

        var actions = new List<PurchaseRequestDetailAction>();
        var labels = new Dictionary<PurchaseRequestDetailAction, string>(DefaultLabels);

        if (canManage && resolvedScreen == PurchaseRequestDetailScreen.ManagementSubmittedReview
            && status.Equals(ProcurementStatus.Submitted, StringComparison.OrdinalIgnoreCase))
        {
            var urgent = priority.Equals(ProcurementPriority.Urgent, StringComparison.OrdinalIgnoreCase);

            if (urgent)
            {
                actions.Add(PurchaseRequestDetailAction.DirectApprove);
                actions.Add(PurchaseRequestDetailAction.RejectRequest);
            }
            else
            {
                actions.Add(PurchaseRequestDetailAction.StartQuoteProcess);
                labels[PurchaseRequestDetailAction.RejectRequest] = "Talebi Reddet";
                actions.Add(PurchaseRequestDetailAction.RejectRequest);
            }
        }

        var showQuotes = false;
        var showPerQuoteApprove = false;

        if (canManage && resolvedScreen == PurchaseRequestDetailScreen.ManagementQuoteReview
            && status.Equals(ProcurementStatus.ManagementQuoteReview, StringComparison.OrdinalIgnoreCase))
        {
            showQuotes = (talep.Teklifler?.Count ?? 0) > 0;
            showPerQuoteApprove = showQuotes && !talep.YonetimOnayKilitli && !talep.HerhangiKalemOnayli;
            actions.Add(PurchaseRequestDetailAction.RejectEntireRequest);
            actions.Add(PurchaseRequestDetailAction.SendQuotesForRevision);
        }

        return new PurchaseRequestDetailUiState
        {
            RequestId = talep.Id.ToString(),
            Status = status,
            Priority = priority,
            Screen = resolvedScreen,
            VisibleActions = actions,
            ActionLabels = labels,
            ShowQuotesList = showQuotes,
            ShowPerQuoteApproveButtons = showPerQuoteApprove,
            RequiresRejectionReason = true,
            RequiresRevisionNote = true
        };
    }

    public static IReadOnlyList<PurchaseRequestQuoteRow> BuildQuoteRows(
        SatinalmaTalep talep,
        string? role)
    {
        var ui = BuildUiState(talep, role, PurchaseRequestDetailScreen.ManagementQuoteReview);
        if (!ui.ShowPerQuoteApproveButtons)
            return [];

        talep.Teklifler ??= [];
        return talep.Teklifler
            .Where(t => !string.IsNullOrWhiteSpace(t.FirmaAdi)
                        || t.Fiyatlar?.Any(f => f.BirimFiyat > 0) == true)
            .Select(t => new PurchaseRequestQuoteRow
            {
                QuoteId = t.Id.ToString(),
                FirmName = string.IsNullOrWhiteSpace(t.FirmaAdi) ? "—" : t.FirmaAdi,
                CanApprove = ui.ShowPerQuoteApproveButtons
            })
            .ToList();
    }

    public static PurchaseRequestDetailMutation? CreateMutation(
        PurchaseRequestDetailAction action,
        SatinalmaTalep talep,
        string? role,
        string? quoteId = null,
        string? note = null)
    {
        var ui = BuildUiState(talep, role);
        if (!ui.VisibleActions.Contains(action)
            && action != PurchaseRequestDetailAction.ApproveQuote)
            return null;

        if (action == PurchaseRequestDetailAction.ApproveQuote)
        {
            if (string.IsNullOrWhiteSpace(quoteId))
                return null;
            if (!ui.ShowPerQuoteApproveButtons)
                return null;
            return PurchaseRequestDetailMutation.ApproveQuote(quoteId);
        }

        return action switch
        {
            PurchaseRequestDetailAction.DirectApprove => PurchaseRequestDetailMutation.DirectApprove(),
            PurchaseRequestDetailAction.RejectRequest or PurchaseRequestDetailAction.RejectEntireRequest
                => PurchaseRequestDetailMutation.Reject(note ?? ""),
            PurchaseRequestDetailAction.StartQuoteProcess => PurchaseRequestDetailMutation.StartQuoteProcess(),
            PurchaseRequestDetailAction.SendQuotesForRevision
                => PurchaseRequestDetailMutation.SendForRevision(note ?? ""),
            _ => null
        };
    }

    public static bool CanManagementDecide(string? role)
    {
        var normalized = KullaniciRolleri.Normalize(role);
        return normalized is KullaniciRolleri.Admin or KullaniciRolleri.Yonetim;
    }
}

File: SatinalmaPro.Android/app/src/main/java/com/satinalmapro/shared/filter/detail/PurchaseRequestDetailPresenter.kt
package com.satinalmapro.shared.filter.detail

import com.satinalmapro.android.core.model.TalepItem
import com.satinalmapro.android.core.roles.KullaniciRolleri
import com.satinalmapro.shared.filter.ProcurementPriority
import com.satinalmapro.shared.filter.ProcurementStatus
import com.satinalmapro.shared.filter.resolvedEnterprisePriority
import com.satinalmapro.shared.filter.resolvedEnterpriseStatus

/**
 * Talep detay — durum, öncelik ve role göre buton görünürlüğü (Android + ortak kurallar).
 */
object PurchaseRequestDetailPresenter {

    private val defaultLabels = mapOf(
        PurchaseRequestDetailAction.DIRECT_APPROVE to "Direkt Onay Ver",
        PurchaseRequestDetailAction.REJECT_REQUEST to "Talebi Reddet",
        PurchaseRequestDetailAction.START_QUOTE_PROCESS to "Teklif Sürecini Başlat",
        PurchaseRequestDetailAction.APPROVE_QUOTE to "Bu Firmayı Onayla",
        PurchaseRequestDetailAction.REJECT_ENTIRE_REQUEST to "Talebi Komple Reddet",
        PurchaseRequestDetailAction.SEND_QUOTES_FOR_REVISION to "Teklifleri Revizeye Gönder"
    )

    fun resolveScreen(talep: TalepItem): PurchaseRequestDetailScreen =
        if (talep.resolvedEnterpriseStatus().equals(ProcurementStatus.MANAGEMENT_QUOTE_REVIEW, ignoreCase = true))
            PurchaseRequestDetailScreen.MANAGEMENT_QUOTE_REVIEW
        else
            PurchaseRequestDetailScreen.MANAGEMENT_SUBMITTED_REVIEW

    fun buildUiState(
        talep: TalepItem,
        role: String?,
        screen: PurchaseRequestDetailScreen? = null
    ): PurchaseRequestDetailUiState {
        val resolvedScreen = screen ?: resolveScreen(talep)
        val status = talep.resolvedEnterpriseStatus()
        val priority = talep.resolvedEnterprisePriority()
        val canManage = canManagementDecide(role)

        val actions = mutableListOf<PurchaseRequestDetailAction>()
        val labels = defaultLabels.toMutableMap()

        if (canManage &&
            resolvedScreen == PurchaseRequestDetailScreen.MANAGEMENT_SUBMITTED_REVIEW &&
            status.equals(ProcurementStatus.SUBMITTED, ignoreCase = true)
        ) {
            val urgent = priority.equals(ProcurementPriority.URGENT, ignoreCase = true)
            if (urgent) {
                actions += PurchaseRequestDetailAction.DIRECT_APPROVE
                actions += PurchaseRequestDetailAction.REJECT_REQUEST
            } else {
                actions += PurchaseRequestDetailAction.START_QUOTE_PROCESS
                actions += PurchaseRequestDetailAction.REJECT_REQUEST
            }
        }

        var showQuotes = false
        var showPerQuoteApprove = false

        if (canManage &&
            resolvedScreen == PurchaseRequestDetailScreen.MANAGEMENT_QUOTE_REVIEW &&
            status.equals(ProcurementStatus.MANAGEMENT_QUOTE_REVIEW, ignoreCase = true)
        ) {
            showQuotes = talep.teklifler.isNotEmpty()
            showPerQuoteApprove = showQuotes && !talep.yonetimOnayKilitli && !talep.herhangiKalemOnayli
            actions += PurchaseRequestDetailAction.REJECT_ENTIRE_REQUEST
            actions += PurchaseRequestDetailAction.SEND_QUOTES_FOR_REVISION
        }

        return PurchaseRequestDetailUiState(
            requestId = talep.id,
            status = status,
            priority = priority,
            screen = resolvedScreen,
            visibleActions = actions,
            actionLabels = labels,
            showQuotesList = showQuotes,
            showPerQuoteApproveButtons = showPerQuoteApprove
        )
    }

    fun buildQuoteRows(talep: TalepItem, role: String?): List<PurchaseRequestQuoteRow> {
        val ui = buildUiState(talep, role, PurchaseRequestDetailScreen.MANAGEMENT_QUOTE_REVIEW)
        if (!ui.showPerQuoteApproveButtons) return emptyList()

        return talep.teklifler
            .filter { it.firmaAdi.isNotBlank() || it.fiyatlar.any { f -> f.birimFiyat > 0 } }
            .map {
                PurchaseRequestQuoteRow(
                    quoteId = it.id,
                    firmName = it.firmaAdi.ifBlank { "—" },
                    canApprove = ui.showPerQuoteApproveButtons
                )
            }
    }

    fun createMutation(
        action: PurchaseRequestDetailAction,
        talep: TalepItem,
        role: String?,
        quoteId: String? = null,
        note: String? = null
    ): PurchaseRequestDetailMutation? {
        val ui = buildUiState(talep, role)

        if (action == PurchaseRequestDetailAction.APPROVE_QUOTE) {
            if (quoteId.isNullOrBlank() || !ui.showPerQuoteApproveButtons) return null
            return PurchaseRequestDetailMutation.approveQuote(quoteId)
        }

        if (action !in ui.visibleActions) return null

        return when (action) {
            PurchaseRequestDetailAction.DIRECT_APPROVE -> PurchaseRequestDetailMutation.directApprove()
            PurchaseRequestDetailAction.REJECT_REQUEST,
            PurchaseRequestDetailAction.REJECT_ENTIRE_REQUEST -> PurchaseRequestDetailMutation.reject(note.orEmpty())
            PurchaseRequestDetailAction.START_QUOTE_PROCESS -> PurchaseRequestDetailMutation.startQuoteProcess()
            PurchaseRequestDetailAction.SEND_QUOTES_FOR_REVISION -> PurchaseRequestDetailMutation.sendForRevision(note.orEmpty())
            else -> null
        }
    }

    fun canManagementDecide(role: String?): Boolean =
        KullaniciRolleri.canManagementDecide(role)
}

private val TalepItem.herhangiKalemOnayli: Boolean
    get() = kalemler.any { !it.onaylananTeklifId.isNullOrBlank() }
